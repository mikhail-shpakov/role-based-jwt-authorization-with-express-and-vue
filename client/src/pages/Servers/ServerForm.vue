<template lang="pug">
.container
  .title.is-size-4 {{ selected.id ? 'Редактировать' : 'Новый сервер' }}
  form.box(
    @submit.prevent="saveChanges"
  )
    template(v-if="selected.id")
      p
        strong ID:
        |  {{ selected.id }}

      .divider

    b-field(label="Название сервера")
      b-input(
        v-model.trim="local.serverName"
        expanded
        required
      )
    b-field.input-margin(label="Тип сервера")
      b-select(
        v-model.trim='local.serverType'
        expanded
      )
        option(
          v-for="option in ['dedicated', 'vds', 'hosting']"
          :key="option"
          :value="option"
        ) {{ option }}

    .buttons.is-pulled-right.button-custom
      b-button(
        v-if="selected.id"
        ref="button-delete"
        type="is-danger"
        icon-left="delete"
        :loading="isLoadingDelete"
        @click.native="confirmDelete()"
      ) Удалить
      b-button(
        ref="button-ed"
        type="is-success"
        :loading="isLoadingChanges"
        :disabled="isCheckStatusButton"
        @click.native="saveChanges(selected.id ? 'edit' : 'add')"
      ) {{ selected.id ? 'Сохранить изменения' : 'Добавить сервер' }}
</template>

<script>
import { mapActions } from 'vuex'

export default {
  name: 'serverForm',
  props: {
    selected: {
      type: Object,
      required: true
    }
  },
  data () {
    return {
      isLoadingChanges: false,
      isLoadingDelete: false,
      local: {
        id: '',
        serverName: '',
        serverType: ''
      }
    }
  },
  methods: {
    ...mapActions('server', [
      'EDIT_SERVER',
      'ADD_SERVER',
      'DELETE_SERVER'
    ]),
    async saveChanges (action) {
      if (this.isCheckStatusButton) return

      this.isLoadingChanges = true

      action === 'edit'
        ? await this.EDIT_SERVER(this.local)
        : await this.ADD_SERVER({ serverName: this.local.serverName, serverType: this.local.serverType })

      this.isLoadingChanges = false
    },
    confirmDelete () {
      this.$buefy.dialog.confirm({
        message: `Удалить сервер с названием <strong>${this.local.serverName}</strong>?`,
        cancelText: 'Отмена',
        confirmText: 'Удалить',
        type: 'is-danger',
        onConfirm: async () => { await this.deleteServer() }
      })
    },
    async deleteServer () {
      this.isLoadingDelete = true
      await this.DELETE_SERVER(this.local.id)
      this.isLoadingDelete = false
    }
  },
  computed: {
    isCheckStatusButton () {
      /**
       * Чтобы кнопка сохранения изменений стала активной,
       * поля в форме должны быть не пустыми
       * и отличаться от уже установленных значений
       */
      if (this.local.serverName.length < 3) return true

      const checkDifValue = Object.keys(this.local).some(v =>
        this.local[v] !== this.selected[v]
      )
      return !checkDifValue
    }
  },
  watch: {
    selected: {
      handler () {
        /**
         * Чтобы не менять props напрямую,
         * перезаписываем значения в $data.local
         */
        this.local = Object.assign({}, this.selected)
      },
      immediate: true
    }
  }
}
</script>

<style lang="sass" scoped>
.input-margin
  margin-top: 15px

form.box
  border: 1px solid $color-alt
  border-radius: 8px
  padding: 20px 10px 110px
  box-shadow: none
  @media (min-width: $display-bp-mobile)
    padding: 20px 20px 70px

  .divider
    border-bottom: 1px solid $color-alt
    margin: 15px -10px
    @media (min-width: $display-bp-mobile)
      margin: 15px -20px

  .button-custom
    margin-top: 15px
</style>

<style lang="sass">
input.input
  box-shadow: none
</style>
