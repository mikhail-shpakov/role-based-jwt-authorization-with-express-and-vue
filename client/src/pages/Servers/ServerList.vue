<template lang="pug">
.container
  p.title.is-size-4 Список серверов
  p.subtitle {{ subtitleText }}

  a(
    v-if="role === 'admin'"
    @click="addServer()"
  ) Добавить сервер
  .table-box
    b-table(
      :data="serverList"
      :columns="columns"
      :loading="isLoading"
      :selected.sync="selected"
      hoverable
      striped
    )
      template(slot='footer')
        .has-text-right Серверов всего: {{ serverList.length }}

      template(slot='empty')
          section.section
            .content.has-text-grey.has-text-centered
              p
                b-icon(
                  icon='emoticon-sad'
                  size='is-large'
                )
              p Список серверов пуст
</template>

<script>
import { mapActions, mapState } from 'vuex'

export default {
  name: 'serverList',
  data () {
    return {
      isLoading: true,
      selected: {},
      columns: [
        {
          field: 'id',
          label: 'ID',
          width: '60'
        },
        {
          field: 'serverName',
          label: 'Название сервера'
        },
        {
          field: 'serverType',
          label: 'Тип сервера'
        }
      ]
    }
  },
  computed: {
    ...mapState('server', {
      serverList: state => state.serverList
    }),
    ...mapState('user', {
      role: state => Object.keys(state.userInfo).length ? state.userInfo.permissions[0] : 'user'
    }),
    subtitleText () {
      return this.role === 'admin'
        ? 'Нажмите на любой сервер, чтобы начать редактирование и Esc, чтобы выйти из режима редактирования'
        : 'Пользователь может только просматривать список серверов'
    }
  },
  methods: {
    ...mapActions('server', [
      'GET_SERVER_LIST'
    ]),
    async fetchServerList () {
      await this.GET_SERVER_LIST()
      this.isLoading = false
    },
    addListenerEsc (e) {
      if (e.code === 'Escape') this.selected = {}
    },
    addServer () {
      const emptyData = {
        id: 0,
        serverName: '',
        serverType: 'dedicated'
      }
      this.$emit('selectServer', emptyData)
    }
  },
  watch: {
    selected () {
      if (this.role === 'admin') {
        this.$emit('selectServer', this.selected)
      }
    },
    serverList: {
      handler () {
        this.selected = {}
      },
      deep: true
    }
  },
  async mounted () {
    await this.fetchServerList()
    document.addEventListener('keyup', e => this.addListenerEsc(e))
  },
  beforeDestroy () {
    document.removeEventListener('keyup', e => this.addListenerEsc(e))
  }
}
</script>

<style lang="sass" scoped>
div.table-box
  border: 1px solid $color-alt
  border-radius: 8px
  margin-top: 10px
  padding: 20px 10px 15px
  @media (min-width: $display-bp-mobile)
    padding: 22px 20px 20px

p.subtitle
  color: $color-alt
  font-size: 15px
  margin: -1rem 0 1.5rem !important
</style>

<style lang="sass">
@media (max-width: $display-bp-mobile)
  div.b-table .has-mobile-cards table.table tr
    box-shadow: none
    border-radius: 6px
    border: 1px solid $color-alt
    margin-top: 10px
    &.table-footer
      border: none
</style>
