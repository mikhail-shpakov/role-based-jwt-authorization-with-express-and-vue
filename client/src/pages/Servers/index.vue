<template lang="pug">
.columns
  .column.is-half
    server-list(
      @selectServer="selectServer"
    )
  .column.form
    transition(
      name="fade"
      mode="out-in"
      :duration="{ enter: 150, leave: 150 }"
    )
      div(
        v-if="!Object.keys(selected).length"
      )
        .img(
          :class="{ 'choice': role === 'admin', 'not-allowed': role === 'user' }"
        )
        p.has-text-centered {{ imgSubtitle }}
      server-form(
        :selected="selected"
        v-else
      )
</template>

<script>
// eslint-disable-next-line no-unused-vars
import methods from '@/services/api/methods'
import { mapState } from 'vuex'
import ServerList from './ServerList'
import ServerForm from './ServerForm'

export default {
  name: 'Servers',
  components: {
    ServerList,
    ServerForm
  },
  data () {
    return {
      selected: {}
    }
  },
  computed: {
    ...mapState('user', {
      role: state => Object.keys(state.userInfo).length ? state.userInfo.permissions[0] : 'user'
    }),
    imgSubtitle () {
      return this.role === 'admin'
        ? 'Сервер для редактирования не выбран'
        : 'Добавлять, удалять и редактировать сервера может только администратор'
    }
  },
  methods: {
    selectServer (value) {
      this.selected = value
    }
  }
}
</script>

<style lang="sass" scoped>
div.columns
  margin: 0

  .column.form
    @media (min-width: $display-bp-desktop)
      padding-left: 50px
      display: flex
      justify-content: center

    div.img
      background-size: 100% 100%
      min-height: 350px
      &.not-allowed
        background-image: url("../../assets/not-allowed.svg")
        margin: auto
        max-width: 240px
      &.choice
        background-image: url("../../assets/choice.svg")
        @media (min-width: $display-bp-desktop)
          min-width: 330px
</style>
