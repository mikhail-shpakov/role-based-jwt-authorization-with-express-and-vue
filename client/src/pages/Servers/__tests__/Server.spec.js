import { createLocalVue, shallowMount } from '@vue/test-utils'
import Vuex from 'vuex'
import Buefy from 'buefy'
import Servers from '../index'
import ServerList from '../ServerList'
import ServerForm from '../ServerForm'

const localVue = createLocalVue()
localVue.use(Buefy)
localVue.use(Vuex)

describe('Servers', () => {
  let store

  beforeEach(() => {
    const state = {
      userInfo: {
        permissions: ['admin']
      }
    }
    store = new Vuex.Store({
      modules: {
        user: {
          state,
          namespaced: true
        }
      }
    })
  })

  it('when load page and user logged in as user, Servers renders correctly', () => {
    store.state.user.userInfo.permissions = ['user']

    const wrapper = shallowMount(Servers, {
      localVue,
      store
    })
    const img = wrapper.find('.img')
    const imgSubtitle = wrapper.find('.has-text-centered').text()

    expect(img.classes('not-allowed')).toBe(true)
    expect(imgSubtitle).toEqual('Добавлять, удалять и редактировать сервера может только администратор')
    expect(wrapper).toMatchSnapshot()
  })

  it('when load page and user logged in as admin, Servers renders correctly', () => {
    const wrapper = shallowMount(Servers, {
      localVue,
      store
    })
    const img = wrapper.find('.img')
    const imgSubtitle = wrapper.find('.has-text-centered').text()

    expect(img.classes('choice')).toBe(true)
    expect(imgSubtitle).toEqual('Сервер для редактирования не выбран')
    expect(wrapper).toMatchSnapshot()
  })

  it('when ServerList emitted event, ServerForm rendering', async () => {
    const wrapper = shallowMount(Servers, {
      localVue,
      store
    })
    const selected = {
      id: 1,
      serverName: 'Супер сервер',
      serverType: 'dedicated'
    }
    wrapper.find(ServerList).vm.$emit('selectServer', selected)

    await wrapper.vm.$nextTick()

    expect(wrapper.contains(ServerForm)).toBe(true)
    expect(wrapper).toMatchSnapshot()
  })
})
