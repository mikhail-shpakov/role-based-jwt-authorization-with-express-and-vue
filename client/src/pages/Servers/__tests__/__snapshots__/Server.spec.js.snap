// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Servers when ServerList emitted event, ServerForm rendering 1`] = `
<div class="columns">
  <div class="column is-half">
    <server-list-stub></server-list-stub>
  </div>
  <div class="column form">
    <transition-stub name="fade" mode="out-in" duration="[object Object]">
      <server-form-stub selected="[object Object]"></server-form-stub>
    </transition-stub>
  </div>
</div>
`;

exports[`Servers when load page and user logged in as admin, Servers renders correctly 1`] = `
<div class="columns">
  <div class="column is-half">
    <server-list-stub></server-list-stub>
  </div>
  <div class="column form">
    <transition-stub name="fade" mode="out-in" duration="[object Object]">
      <div>
        <div class="img choice"></div>
        <p class="has-text-centered">Сервер для редактирования не выбран</p>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`Servers when load page and user logged in as user, Servers renders correctly 1`] = `
<div class="columns">
  <div class="column is-half">
    <server-list-stub></server-list-stub>
  </div>
  <div class="column form">
    <transition-stub name="fade" mode="out-in" duration="[object Object]">
      <div>
        <div class="img not-allowed"></div>
        <p class="has-text-centered">Добавлять, удалять и редактировать сервера может только администратор</p>
      </div>
    </transition-stub>
  </div>
</div>
`;
